src/ray/raylet/node_manager.cc

```diff
@@ -1132,63 +1068,64 @@ void NodeManager::ProcessWaitRequestMessage(
     }
   }
 
-  const TaskID &current_task_id = from_flatbuf<TaskID>(*message->task_id());
+  const TaskID &current_task_id = TaskID::FromBinary(request.task_id());
+  const WorkerID &worker_id = WorkerID::FromBinary(request.worker_id());
   bool client_blocked = !required_object_ids.empty();
   if (client_blocked) {
-    HandleTaskBlocked(client, required_object_ids, current_task_id);
+    HandleTaskBlocked(worker_id, required_object_ids, current_task_id);
   }
 
   ray::Status status = object_manager_.Wait(
       object_ids, wait_ms, num_required_objects, wait_local,
-      [this, client_blocked, client, current_task_id](std::vector<ObjectID> found,
-                                                      std::vector<ObjectID> remaining) {
-        // Write the data.
-        flatbuffers::FlatBufferBuilder fbb;
-        flatbuffers::Offset<protocol::WaitReply> wait_reply = protocol::CreateWaitReply(
-            fbb, to_flatbuf(fbb, found), to_flatbuf(fbb, remaining));
-        fbb.Finish(wait_reply);
-
-        auto status =
-            client->WriteMessage(static_cast<int64_t>(protocol::MessageType::WaitReply),
-                                 fbb.GetSize(), fbb.GetBufferPointer());
-        if (status.ok()) {
-          // The client is unblocked now because the wait call has returned.
-          if (client_blocked) {
-            HandleTaskUnblocked(client, current_task_id);
-          }
-        } else {
-          // We failed to write to the client, so disconnect the client.
-          RAY_LOG(WARNING)
-              << "Failed to send WaitReply to client, so disconnecting client";
-          // We failed to send the reply to the client, so disconnect the worker.
-          ProcessDisconnectClientMessage(client);
-        }
+      [this, client_blocked, worker_id, current_task_id, reply, send_reply_callback](
+          std::vector<ObjectID> found, std::vector<ObjectID> remaining) {
+        IdVectorToProtobuf<ObjectID, rpc::WaitReply>(found, *reply,
+                                                     &rpc::WaitReply::add_found);
+        IdVectorToProtobuf<ObjectID, rpc::WaitReply>(remaining, *reply,
+                                                     &rpc::WaitReply::add_remaining);
+
+        // Send reply to finish this wait request.
+        send_reply_callback(
+            Status::OK(),
+            /* success callback */
+            [this, client_blocked, worker_id, current_task_id]() {
+              if (client_blocked) {
+                HandleTaskUnblocked(worker_id, current_task_id);
+              }
+            },
+            /* failure callback */
+            [this, worker_id]() { ProcessDisconnectClientMessage(worker_id); });
```

Comment:
It seems like this isn't necessary anymore since all worker failures will be handled with heartbeats. Let's just use `nullptr` instead to be consistent with the other worker failures. Another option is we *always* call `ProcessDisconnectClientMessage` if a reply fails.